src/S3SparseIterator.cpp

```diff
@@ -87,30 +88,32 @@ const void* S3SparseIterator::get_next_fast() {
 
 #ifdef DEBUG
   if (ret.second != -1) {
-    std::cout << "get_next_fast::ret.second: " << ret.second << std::endl;
+    std::cout << "getNext::ret.second: " << ret.second << std::endl;
   }
 #endif
 
   to_delete = ret.second;
 
   if (num_minibatches_ready < 200 && pref_sem.getvalue() < (int)read_ahead) {
 #ifdef DEBUG
-    std::cout << "get_next_fast::pref_sem.signal" << std::endl;
+    std::cout << "getNext::pref_sem.signal" << std::endl;
 #endif
     pref_sem.signal();
   }
 
-  return ret.first;
+  return std::make_shared<SparseDataset>(reinterpret_cast<const char*>(ret.first),
```

Comment:
Before returning, I think it would be good to run a check() on the dataset to make sure it is valid.